site/_posts/2020-09-01-microservices-gateway.md

```diff
@@ -0,0 +1,267 @@
+
+Clone the [project](https://github.com/FusionAuth/fusionauth-example-node-services-gateway) onto your local computer and `cd` into the directory.
+
+You'll notice three folders corresponding to our applications: `gateway`, `product-catalog`, and `product-inventory`. Go ahead and `cd` into the `product-catalog` application and do the following:
+
+* Run `npm install` to install the dependencies.
+* Start the application by running `npm start`. It should be running on port `3001`, which is defined in `bin/www`.
+
+Now that your application is up and running, you should be able to send a request to it and get a response. Run this `curl` command in a separate terminal window and you should get a successful response with an empty list of products (`products: []`):
+
+```shell
+curl http://localhost:3001/products
+```
+
+Let's open up the hood on this and check out the `routes/index.js` file that gave us our `/products` route.
+
+```javascript
+const express = require('express');
+const router = express.Router();
+
+router.get('/products', function(req, res, next) {
+  res.json('products: []')
+});
+
+router.get('/products/:id', function(req, res, next) {
+  res.json(`product: ${req.params.id}`)
+});
+
+module.exports = router;
+```
+
+We've created two basic routes, `/products` and `/products/:id`, so we can get a list of products and a single product. Obviously, for a real microservice, these routes would request the product information from a datastore. For now, the former is returning an empty array `[]` and the latter returns the product id requested.
+
+Try modifying your curl request to add a product ID, and notice that the response will indicate the specific ID you requested.
+
+The Product Catalog service is ready to go!
+
+## Product Inventory service
+
+Open up another terminal window and enter the `product-inventory` folder. Run `npm install` to install needed dependencies.
+
+Here's what our Product Inventory service looks like (in `routes/index.js`):
+
+```javascript
+const express = require('express');
+const router = express.Router();
+
+router.get('/branches/:id/products', function(req, res, next) {
+  const roles = req.headers.roles;
+  if (roles && roles.includes('admin')) {
+    res.json(`Products for branch #${req.params.id}`);
+  } else {
+    res.redirect(403, 'http://localhost:3000');
+    return;
+  }
+});
+
+module.exports = router;
+```
+
+In this service, we've just got one route, to get products for a specific store or branch. Notice, however, that we're allowing access (or denying it) based on the inclusion of an `admin` role in the `roles` header. The API gateway application will be responsible for passing this data to our Product Inventory service.
+
+If you were to start the service--go ahead and do so with `npm start`--and send a request to `http://localhost:3002/branches/1/products`, you should receive a 403. You can simulate a successful response by adding a `roles` header with a value of `admin`:
```

Comment:
These hyphens tripped me up a bit. Parenthesis might read cleaner: `If you were to start the service (go ahead and do so with npm start) and send`